### PR TITLE
Fix #7497: Rate App News Card & Logic

### LINKED ISSUE #7497: Rate app news card
https://github.com/brave/internal/issues/1023

### PR DESCRIPTION
**Pull request by soner-yuksel**
## Summary of Changes

This pull request fixes #7497

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Test Plan:
 Test Display rules and actions mentioned in ticket

## Screenshots:

res-tested1 3 1 3

https://github.com/brave/brave-ios/assets/6643505/faf04b39-2064-44d7-8cd7-dcb9d7445741

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift [hunk 1]
```diff
@@ -424,12 +440,53 @@ class BraveNewsSectionProvider: NSObject, NTPObservableSectionProvider {
     }
   }
 
-  private func contextMenu(from feedList: @escaping (Int) -> FeedItem, card: FeedCard, indexPath: IndexPath) -> FeedItemMenu {
+  private func contextMenu(from feedList: @escaping (Int) -> FeedItem, card: FeedCard, indexPath: IndexPath, isRatingCardPair: Bool = false) -> FeedItemMenu {
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift]: There's no reason to add this logic to the main `contextMenu(from:card:indexPath:)`, since its only called when you're setting up the rating card. Move the menu generation for this card to its own method

File: Sources/BraveNews/Cards/FeedHeadlineViews.swift [hunk 2]
```diff
@@ -64,6 +65,59 @@ public class SmallHeadlineCardView: HeadlineCardView {
   }
 }
 
+public class SmallRateCardView: FeedCardBackgroundButton, FeedCardContent {
```
Comment (kylehickinson) [inline Sources/BraveNews/Cards/FeedHeadlineViews.swift]: Lets move these to their own files

File: Sources/Growth/AppReviewManager.swift [hunk 3]
```diff
@@ -146,6 +156,30 @@ public class AppReviewManager: ObservableObject {
...
+      // First presentation Date, no recorded presentation
+      Preferences.Review.newsCardShownDate.value = Date()
+      return true
+    }
+    
+    if abs(daysSinceLastRequest) < Constants.minDaysBetweenRatingCardPresented {
+      return false
+    }
+    
+    Preferences.Review.newsCardShownDate.value = Date()
```
Comment (kylehickinson) [inline Sources/Growth/AppReviewManager.swift]: This isn't true though? A card may be generated but the user may not scroll down and see it

File: Sources/BraveNews/Composer/FeedDataSource.swift [hunk 4]
```diff
@@ -854,18 +855,12 @@ public class FeedDataSource: ObservableObject {
...
-      .sponsor,
-      .fillUsing(
-        FilteredFillStrategy(isIncluded: { $0.source.category == Self.topNewsCategory }),
-        fallback: DefaultFillStrategy(),
-        [
-          .headline(paired: false)
-        ]),
-      .braveAd,
-      .repeating([
+    func createRepatingFeedElements(withRatingCard: Bool) -> [FeedSequenceElement] {
```
Comment (kylehickinson) [inline Sources/BraveNews/Composer/FeedDataSource.swift]: typo: `createRepatingFeedElements` → `createRepeatingFeedElements`

File: Sources/BraveNews/FeedCardContent.swift [hunk 5]
```diff
@@ -37,4 +37,8 @@ public enum FeedItemAction: Equatable {
   case opened(inNewTab: Bool = false, switchingToPrivateMode: Bool = false)
   /// Toggle the enabled status for the source of the feed item
   case toggledSource
+  /// The user choose to navigate Appstore for app rating
+  case pressedAppRatingCard
+  /// The user choose to hide card for app rating
+  case hiddenAppRatingCard
```
Comment (kylehickinson) [inline Sources/BraveNews/FeedCardContent.swift]: Lets remove these from here and make the `actionHandler` just an unused variable on the ratings card. We can have a separate action setup for the ratings card instead

File: Sources/BraveNews/FeedItemView.swift [hunk 6]
```diff
@@ -365,6 +365,46 @@ extension FeedItemView {
...
+        children: [
+          .thumbnail(.aspectRatio(1.5)),
+          .stack(
+            .init(
+              axis: .vertical,
+              spacing: 4,
+              padding: UIEdgeInsets(top: 12, left: 12, bottom: 12, right: 12),
+              children: [
+                .title(.init(numberOfLines: 2, font: .systemFont(ofSize: 24.0, weight: .bold))),
+                .description(.init(numberOfLines: 3, font: .systemFont(ofSize: 16.0, weight: .semibold))),
```
Comment (kylehickinson) [inline Sources/BraveNews/FeedItemView.swift]: These font sizes are too big and look odd next to the regular card

File: Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift [hunk 7]
```diff
@@ -70,6 +72,8 @@ class BraveNewsSectionProvider: NSObject, NTPObservableSectionProvider {
     collectionView.register(FeedCardCell<BraveNewsEmptyFeedView>.self)
     collectionView.register(FeedCardCell<HeadlineCardView>.self)
     collectionView.register(FeedCardCell<SmallHeadlinePairCardView>.self)
+    collectionView.register(FeedCardCell<SmallRateCardView>.self)
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift]: This card doesn't need to be registered since its only shown inside of `SmallHeadlineRatePairCardView`

File: Sources/BraveNews/FeedCardContent.swift (hunk 5 above)
Comment (kylehickinson) [reply Sources/BraveNews/FeedCardContent.swift]: i.e. 

```swift
enum RatingCardAction {
  case rateBrave
  case hideCard
}
```

File: Sources/BraveNews/FeedItemView.swift [hunk 8]
```diff
@@ -365,6 +365,46 @@ extension FeedItemView {
...
+    /// ├─────────────────┤
+    /// │ action button   │
+    /// ╰─────────────────╯
+    /// ```
+    ///
+    static let rateCard = Layout(
+      root: .init(
+        axis: .vertical,
+        children: [
+          .thumbnail(.aspectRatio(1.5)),
```
Comment (kylehickinson) [inline Sources/BraveNews/FeedItemView.swift]: Can we also fix the thumbnail? In the screenshots/video its cutoff

**Review by kylehickinson — CHANGES_REQUESTED** [review]
Nice! 🙂

File: Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift (hunk 1 above)
Comment (soner-yuksel) [reply Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift]: Moved to a new method at https://github.com/brave/brave-ios/pull/7589/commits/a4f446ca688845f1e0b47127a94c5b28f4e09231

File: Sources/BraveNews/Cards/FeedHeadlineViews.swift (hunk 2 above)
Comment (soner-yuksel) [reply Sources/BraveNews/Cards/FeedHeadlineViews.swift]: Fix at https://github.com/brave/brave-ios/pull/7589/commits/a4f446ca688845f1e0b47127a94c5b28f4e09231

File: Sources/BraveNews/Composer/FeedDataSource.swift (hunk 4 above)
Comment (soner-yuksel) [reply Sources/BraveNews/Composer/FeedDataSource.swift]: Fix at https://github.com/brave/brave-ios/pull/7589/commits/a4f446ca688845f1e0b47127a94c5b28f4e09231

File: Sources/BraveNews/FeedCardContent.swift (hunk 5 above)
Comment (soner-yuksel) [reply Sources/BraveNews/FeedCardContent.swift]: Created a new enum at https://github.com/brave/brave-ios/pull/7589/commits/a4f446ca688845f1e0b47127a94c5b28f4e09231

File: Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift (hunk 7 above)
Comment (soner-yuksel) [reply Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift]: Fix at https://github.com/brave/brave-ios/pull/7589/commits/a4f446ca688845f1e0b47127a94c5b28f4e09231

File: Sources/BraveNews/FeedItemView.swift (hunk 6 above)
Comment (soner-yuksel) [reply Sources/BraveNews/FeedItemView.swift]: Fix at https://github.com/brave/brave-ios/pull/7589/commits/8f5682712a33dc7f60d1c39b87067dc65c51458e

File: Sources/BraveNews/FeedItemView.swift (hunk 8 above)
Comment (soner-yuksel) [reply Sources/BraveNews/FeedItemView.swift]: Design has updated using Android Design Fix at https://github.com/brave/brave-ios/pull/7589/commits/8f5682712a33dc7f60d1c39b87067dc65c51458e

**Comment by soner-yuksel** [comment]
The ticket is updated with Reference Design link for Android. And Screenshot in the description is updated along with the change.

File: Sources/Growth/AppReviewManager.swift (hunk 3 above)
Comment (soner-yuksel) [reply Sources/Growth/AppReviewManager.swift]: Thats a great catch, I moved the preference set when the cell is created. Altered at https://github.com/brave/brave-ios/pull/7589/commits/350b2ac81b755972d6afdb631af508545b972454

File: Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift [hunk 9]
```diff
@@ -335,6 +338,14 @@ class BraveNewsSectionProvider: NSObject, NTPObservableSectionProvider {
       cell.content.smallHeadelineCardViews.right.feedView.setupWithItem(pair.second)
       cell.content.actionHandler = handler(from: { $0 == 0 ? pair.first : pair.second }, card: card, indexPath: indexPath)
       cell.content.contextMenu = contextMenu(from: { $0 == 0 ? pair.first : pair.second }, card: card, indexPath: indexPath)
+      // The rating card is presented
+      Preferences.Review.newsCardShownDate.value = Date()
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift]: Technically this is also not guaranteed that the cell is visible anymore 😅 You may need to use `willDisplayCell` instead to guarantee it

Comment (kylehickinson) [reply Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift]: Also this is on the wrong switch case. Its on `headlinePair` instead of `headlineRatingCardPair` below this

File: Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift [hunk 10]
```diff
@@ -375,9 +386,15 @@ class BraveNewsSectionProvider: NSObject, NTPObservableSectionProvider {
     }
   }
 
-  private func handler(from feedList: @escaping (Int) -> FeedItem, card: FeedCard, indexPath: IndexPath) -> (Int, FeedItemAction) -> Void {
+  private func handler(from feedList: @escaping (Int) -> FeedItem?, card: FeedCard, indexPath: IndexPath, isRatingCardPair: Bool = false) -> (Int, FeedItemAction) -> Void {
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift]: Lets separate this out like we did context menu

File: Sources/BraveNews/Cards/RateCardViews.swift [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    smallHeadlineRateCardViews.ratingCard.actionHandler = { [weak self] _, action in
+      self?.actionHandler?(1, action)
+    }
+    
+    smallHeadlineRateCardViews.smallHeadline.contextMenu = FeedItemMenu({ [weak self] _ -> UIMenu? in
+      return self?.contextMenu?.menu?(0)
+    })
+    smallHeadlineRateCardViews.ratingCard.contextMenu = FeedItemMenu({ [weak self] _ -> UIMenu? in
+      return self?.contextMenu?.menu?(1)
+    })
```
Comment (kylehickinson) [inline Sources/BraveNews/Cards/RateCardViews.swift]: See what I meant when I said split these out basically is that I wanted us to just keep the usual `actionHandler` + `contextMenu` setup for the `smallHeadline` here, since it is a news item, but create a _separate_ action handler in this view that doesn't use `actionHandler`/`contextMenu` since they dont make sense for it and use a custom closure to handle it.  You can even just move the entire menu generation here.

Basically I don't want the ratings card to be a hacked feed item, it should just be a separate view/card entirely, having its own actions similar to the onboarding/error card

File: Sources/BraveNews/FeedCardContent.swift [hunk 12]
```diff
@@ -38,3 +38,11 @@ public enum FeedItemAction: Equatable {
...
   case toggledSource
 }
+
+/// The actions you can perform on a rate card item
+public enum RatingCardAction {
+  // The user choose to navigate Appstore for app rating
+  case rateBrave
+  // The user choose to hide card for app rating
+  case hideCard
+}
```
Comment (kylehickinson) [inline Sources/BraveNews/FeedCardContent.swift]: This can probably go in the RatingCardsView file

**Review by kylehickinson — CHANGES_REQUESTED** [review]
The design is _much_ nicer! Can we please also make the button text white instead of blurple?

File: Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift (hunk 9 above)
Comment (soner-yuksel) [reply Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift]: Thanks for the tip it is is moved to `willDisplayCell` Fix at https://github.com/brave/brave-ios/pull/7589/commits/ebbe26fa4af9e4be47f82ac8e078025b0f0f5031

File: Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift (hunk 10 above)
Comment (soner-yuksel) [reply Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift]: Separation is done both for handler and context menu fix at https://github.com/brave/brave-ios/pull/7589/commits/ebbe26fa4af9e4be47f82ac8e078025b0f0f5031

File: Sources/BraveNews/Cards/RateCardViews.swift (hunk 11 above)
Comment (soner-yuksel) [reply Sources/BraveNews/Cards/RateCardViews.swift]: Menu generation is moved custom  action is added, Please look at https://github.com/brave/brave-ios/pull/7589/commits/ebbe26fa4af9e4be47f82ac8e078025b0f0f5031

File: Sources/BraveNews/FeedCardContent.swift (hunk 12 above)
Comment (soner-yuksel) [reply Sources/BraveNews/FeedCardContent.swift]: Placed RatingCardView https://github.com/brave/brave-ios/pull/7589/commits/ebbe26fa4af9e4be47f82ac8e078025b0f0f5031

**Comment by soner-yuksel** [comment]
> The design is _much_ nicer! Can we please also make the button text white instead of blurple?

The button text is retuned to white.

Simulator Screenshot - iPhone 14 - 2023-06-20 at 14 57 48

File: Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift [hunk 13]
```diff
@@ -375,16 +389,23 @@ class BraveNewsSectionProvider: NSObject, NTPObservableSectionProvider {
     }
   }
 
-  private func handler(from feedList: @escaping (Int) -> FeedItem, card: FeedCard, indexPath: IndexPath) -> (Int, FeedItemAction) -> Void {
+  private func handler(from feedList: @escaping (Int) -> FeedItem?, card: FeedCard, indexPath: IndexPath) -> (Int, FeedItemAction) -> Void {
     return { [weak self] index, action in
-      self?.actionHandler(.itemAction(action, context: .init(item: feedList(index), card: card, indexPath: indexPath)))
+      guard let item = feedList(index) else { return }
+      self?.actionHandler(.itemAction(action, context: .init(item: item, card: card, indexPath: indexPath)))
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift]: This can be reverted

File: Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift [hunk 14]
```diff
@@ -375,16 +389,23 @@ class BraveNewsSectionProvider: NSObject, NTPObservableSectionProvider {
...
 
   private func handler(for item: FeedItem, card: FeedCard, indexPath: IndexPath) -> (Int, FeedItemAction) -> Void {
     return handler(from: { _ in item }, card: card, indexPath: indexPath)
   }
 
+  private func rateCardhandler() -> (RatingCardAction) -> Void {
+    return { [weak self] action in
+      self?.actionHandler(.rateCardAction(.rateBrave))
+    }
+  }
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift]: You can probably just inline this where you're setting up the card instead of have a separate function since its small

File: Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift [hunk 15]
```diff
@@ -441,8 +462,9 @@ class BraveNewsSectionProvider: NSObject, NTPObservableSectionProvider {
       itemActionHandler(.toggledSource, context)
     }
 
-    return .init { [weak self] index -> UIMenu? in
+    let newItemAction: FeedItemMenu = .init { [weak self] index -> UIMenu? in
       guard let self = self else { return nil }
+
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift]: Can be reverted along with the part below

File: Sources/BraveNews/Cards/FeedHeadlineViews.swift [hunk 16]
```diff
@@ -5,6 +5,7 @@
 
 import Foundation
 import UIKit
+import BraveStrings
```
Comment (kylehickinson) [inline Sources/BraveNews/Cards/FeedHeadlineViews.swift]: Still needed?

File: Sources/BraveShared/Extensions/UIActionExtensions.swift [hunk 17]
```diff
@@ -14,7 +14,7 @@ extension UIAction {
   ///   1. See a change that happens based on their selection
   ///   2. Ensure that if the view which has shown the context menu were to be removed from the view
   ///      heirarchy as a result of the action, that the context menu does not crash on dismissal.
-  static func deferredActionHandler(_ handler: @escaping UIActionHandler) -> UIActionHandler {
+  public static func deferredActionHandler(_ handler: @escaping UIActionHandler) -> UIActionHandler {
```
Comment (kylehickinson) [inline Sources/BraveShared/Extensions/UIActionExtensions.swift]: Lets move this into `BraveUI` instead of `BraveShared`

**Review by kylehickinson — CHANGES_REQUESTED** [review]
Just a few small things, everything else looks great

File: Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift (hunk 13 above)
Comment (soner-yuksel) [reply Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift]: Reverted at https://github.com/brave/brave-ios/pull/7589/commits/e5ca33debfb25eca4c069bdc2c2670c82603fb64

File: Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift (hunk 14 above)
Comment (soner-yuksel) [reply Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift]: Changed at https://github.com/brave/brave-ios/pull/7589/commits/e5ca33debfb25eca4c069bdc2c2670c82603fb64

File: Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift [hunk 18]
```diff
@@ -336,6 +343,15 @@ class BraveNewsSectionProvider: NSObject, NTPObservableSectionProvider {
       cell.content.actionHandler = handler(from: { $0 == 0 ? pair.first : pair.second }, card: card, indexPath: indexPath)
       cell.content.contextMenu = contextMenu(from: { $0 == 0 ? pair.first : pair.second }, card: card, indexPath: indexPath)
       return cell
+    case .headlineRatingCardPair(let item):
+      let cell = collectionView.dequeueReusableCell(for: indexPath) as FeedCardCell<SmallHeadlineRatePairCardView>
+      cell.content.smallHeadlineRateCardViews.smallHeadline.feedView.setupWithItem(item)
+      cell.content.actionHandler = handler(for: item, card: card, indexPath: indexPath)
+      cell.content.rateCardActionHandler = { [weak self] _ in
+        self?.actionHandler(.rateCardAction(.rateBrave))
+      }
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift]: Small bug here: You aren't passing in the action from the closure arg, so it'll always rate and hide car will not do anything

File: Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift (hunk 15 above)
Comment (soner-yuksel) [reply Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift]: Fixed at https://github.com/brave/brave-ios/pull/7589/commits/e5ca33debfb25eca4c069bdc2c2670c82603fb64

File: Sources/BraveNews/Cards/FeedHeadlineViews.swift (hunk 16 above)
Comment (soner-yuksel) [reply Sources/BraveNews/Cards/FeedHeadlineViews.swift]: Nope reverted at https://github.com/brave/brave-ios/pull/7589/commits/e5ca33debfb25eca4c069bdc2c2670c82603fb64

File: Sources/BraveShared/Extensions/UIActionExtensions.swift (hunk 17 above)
Comment (soner-yuksel) [reply Sources/BraveShared/Extensions/UIActionExtensions.swift]: Placed at https://github.com/brave/brave-ios/pull/7589/commits/e5ca33debfb25eca4c069bdc2c2670c82603fb64

File: Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift (hunk 18 above)
Comment (soner-yuksel) [reply Sources/Brave/Frontend/Browser/New Tab Page/Sections/BraveNewsSectionProvider.swift]: Ouff it is big bug actually :) thanks for the catch change at https://github.com/brave/brave-ios/pull/7589/commits/e97cbdd6712f2dcc77b37a01550ed1a39cf45338

Also a small UI glitch is fixed

**Review by kylehickinson — COMMENTED** [review]
🎉

**Review by kylehickinson — APPROVED** [review]
🎉
